src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -274,14 +314,31 @@ internal AnalyzerOptions AnalyzerOptions
         /// Use <see cref="ExecuteCompilationStartActions(ImmutableArray{CompilationStartAnalyzerAction}, HostCompilationStartAnalysisScope)"/> API
         /// to get execute these actions to get the per-compilation analyzer actions.
         /// </remarks>
-        public void ExecuteInitializeMethod(DiagnosticAnalyzer analyzer, HostSessionStartAnalysisScope sessionScope)
+        public void ExecuteInitializeMethod(
+            DiagnosticAnalyzer analyzer,
+            HostSessionStartAnalysisScope sessionScope)
         {
             var context = new AnalyzerAnalysisContext(analyzer, sessionScope);
 
             // The Initialize method should be run asynchronously in case it is not well behaved, e.g. does not terminate.
             ExecuteAndCatchIfThrows(
                 analyzer,
-                data => data.analyzer.Initialize(data.context),
+                data =>
+                {
+                    // If this analyzer is actually an artifact generator, and the compiler is requesting artifacts be
+                    // generated, then actually initialize the generator with the necessary context values it needs to
+                    // generate artifacts.  In all other circumstances, we will not initialize it, and as such it will
+                    // not have any opportunity to register callbacks for us to call into in the future.
+                    if (analyzer is ArtifactProducer artifactProducer)
+                    {
+                        if (CreateArtifactStream == null)
+                            return;
+
+                        artifactProducer.CreateArtifactStream = CreateArtifactStream;
```

Comment:
this part isa bit ugly, but is as per our meeting.  since this can only run as part of batch compile, it's ok for these to be internally stateful in this manner.

Reply:
Do you want to mention in the comment that we expect this to be only called in batch compile scenarios?

Reply:
sure.